File: frontend/demo/component/button/button-disable-long-action.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/button';
import '@vaadin/horizontal-layout';
import '@vaadin/progress-bar';
import './fake-progress-bar';
import { html, LitElement } from 'lit';
import { customElement, query, state } from 'lit/decorators.js';
import { applyTheme } from 'Frontend/demo/theme';
import type { FakeProgressBar } from './fake-progress-bar';

@customElement('button-disable-long-action')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  @state()
  private isDisabled = false;

  @query('fake-progress-bar')
  private fakeProgressBar!: FakeProgressBar;

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-horizontal-layout theme="spacing" style="align-items: center;">
        <vaadin-button
          ?disabled=${this.isDisabled}
          @click=${() => {
            this.isDisabled = true;
            this.fakeProgressBar.simulateProgress();
          }}
          style="flex: none;"
          >Perform Action</vaadin-button
        >
        <fake-progress-bar
          @progress-end=${() => {
            this.isDisabled = false;
          }}
        ></fake-progress-bar>
      </vaadin-horizontal-layout>
      <!-- end::snippet[] -->
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/button/ButtonDisableLongAction.java
package com.vaadin.demo.component.button;

import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.orderedlayout.FlexComponent;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("button-disable-long-action")
public class ButtonDisableLongAction extends Div {
    public ButtonDisableLongAction() {
        // tag::snippet[]
        Button button = new Button("Perform Action");
        FakeProgressBar progressBar = new FakeProgressBar();
        button.setDisableOnClick(true);
        button.addClickListener(event -> progressBar.simulateProgress());

        progressBar.addProgressEndListener(event -> {
            button.setEnabled(true);
        });
        // end::snippet[]

        button.getStyle().set("flex", "none");
        HorizontalLayout horizontalLayout = new HorizontalLayout(button, progressBar);
        horizontalLayout.setAlignItems(FlexComponent.Alignment.CENTER);
        add(horizontalLayout);
    }

    public static class Exporter extends DemoExporter<ButtonDisableLongAction> { // hidden-source-line
    } // hidden-source-line
}
